text/0000-standardize-error-docs.md

```diff
@@ -0,0 +1,55 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This RFC makes three changes to error docs in order to fulfill this RFC's (and [RFC 1567]'s) goals:
+
+- Firstly, that the style of error docs is not defined by RFC but by a "style spec" located in the `rust-lang/rust` repository. This style spec can be updated from time-to-time in order to ensure consistent formatting across error docs. Any change to the style spec must be with the consensus of the dev-tools team. It is recommended that the style spec is initially extracted from [RFC 1567].
+- Secondly, that the idea of "labels" is introduced to error docs. These would specify certain attributes about an error code that are important for users to know. Error codes can be marked with labels easily and are verified at compile-time (this would probably happen in the `register_diagnostics` macro). These labels would be added as markdown headers (and any other potential places) at doc-time and would replace the current typo-prone system of copying long lines around. Labels are defined in the style spec. Three  label-candidates are exampled here:
+  - `removed`: An error code has been removed, and therefore the labeled docs docs exist for backwards-compatibility purposes only. The docs may also contain a note explaining *where* the error code was moved to, if at all. (The style spec can specify this more fully)
+  - `internal`: An error code is internal to the compiler/standard library. This means that it can only occur when using a perma-unstable feature gate which is explicitly marked as only intended for use internally.
+  - `feature_gated`: An error code can only be emitted when using an unstable feature. This error is subject to change or removal, all the normal feature gate non-guarantees apply.
+- Thirdly, that error docs (including the new labels feature) are aggressively linted according to the style spec. This helps keep the error docs consistently styled.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+It requires a lot of effort:
+- The error docs check in `tidy` will have to be kept up to date with the style spec. 
+- Error docs will have to be kept up to date with the style spec.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Some changes must be made, otherwise PR authors coming up with their own formatting for error docs will continue to increase. *This is already happening* and only makes error docs harder to understand.
```

Comment:
The problem here is that we need a better way to "split" the error code explanation. For example have the following parts well defined:
 1. Short explanation
 2. Erroneous code example
 3. Long explanation of what's going on
 4. (optional) How to fix it
 5. (optional) Fixed code example

If we have a format that forces to follow this, it'd be much simpler.

Reply:
Yes, obviously this gets more complex when you bring in code examples that involve multiple crates, feature gates, etc. Also things like defining where the error is with comments (like UI tests) is a bit nonstandard. 

Hopefully one day I can have some fun with a finite-state machine parser in tidy :)

Reply:
Well, since you're adding labels, we can also add labels for parts. That would make it much simpler.

Reply:
I don't quite get it? Labels are optional attributes for certain error codes (like real Rust attributes) whereas that sounds more mandatory (like trait impls) with more internal compiler maintenance, not just a style spec.

Reply:
The primary goal of this RFC is to enforce a style on the error code explanations/ You suggested to add labels to know the "kind" of the error code. We can go even further and even force to have specific sections that can be checked with `tidy` to prevent having multiple styles (or at least limit it even more).

Reply:
Yeah, as mentioned in the RFC, tidy linting is a important part of this. I'm not sure about expanding labels like this. Not opposed, just need to think about it for a bit.